File: frontend/lib/authentication/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:safe_zone/authentication/services/auth0_service.dart';

/// HTTP client that automatically adds Auth0 JWT token to requests
class AuthenticatedHttpClient extends http.BaseClient {
  AuthenticatedHttpClient({
    required Auth0Service auth0Service,
    http.Client? innerClient,
  })  : _auth0Service = auth0Service,
        _innerClient = innerClient ?? http.Client();

  final Auth0Service _auth0Service;
  final http.Client _innerClient;

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) async {
    // Get the access token
    final accessToken = await _auth0Service.getAccessToken();

    // Add Authorization header if token exists
    if (accessToken != null && accessToken.isNotEmpty) {
      request.headers['Authorization'] = 'Bearer $accessToken';
    }

    // Add Content-Type header for JSON requests
    if (!request.headers.containsKey('Content-Type') &&
        (request.method == 'POST' ||
            request.method == 'PUT' ||
            request.method == 'PATCH')) {
      request.headers['Content-Type'] = 'application/json';
    }

    return _innerClient.send(request);
  }

  @override
  void close() {
    _innerClient.close();
    super.close();
  }
}

/// Helper class for making authenticated API calls
class ApiClient {
  ApiClient({
    required Auth0Service auth0Service,
    required String baseUrl,
  })  : _auth0Service = auth0Service,
        _baseUrl = baseUrl,
        _client = AuthenticatedHttpClient(auth0Service: auth0Service);

  final Auth0Service _auth0Service;
  final String _baseUrl;
  final AuthenticatedHttpClient _client;

  /// Make a GET request
  Future<http.Response> get(
    String path, {
    Map<String, String>? headers,
  }) async {
    final url = Uri.parse('$_baseUrl$path');
    return _client.get(url, headers: headers);
  }

  /// Make a POST request
  Future<http.Response> post(
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? headers,
  }) async {
    final url = Uri.parse('$_baseUrl$path');
    return _client.post(
      url,
      body: body != null ? jsonEncode(body) : null,
      headers: headers,
    );
  }

  /// Make a PUT request
  Future<http.Response> put(
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? headers,
  }) async {
    final url = Uri.parse('$_baseUrl$path');
    return _client.put(
      url,
      body: body != null ? jsonEncode(body) : null,
      headers: headers,
    );
  }

  /// Make a PATCH request
  Future<http.Response> patch(
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? headers,
  }) async {
    final url = Uri.parse('$_baseUrl$path');
    return _client.patch(
      url,
      body: body != null ? jsonEncode(body) : null,
      headers: headers,
    );
  }

  /// Make a DELETE request
  Future<http.Response> delete(
    String path, {
    Map<String, String>? headers,
  }) async {
    final url = Uri.parse('$_baseUrl$path');
    return _client.delete(url, headers: headers);
  }

  /// Close the client
  void close() {
    _client.close();
  }
}

File: web/safezone-alerts/src/lib/api-client.ts
/**
 * Authenticated fetch utility for making API requests with credentials
 * Automatically includes authentication headers and handles CORS with credentials
 */

import { API_CONFIG } from './api-config';

interface FetchOptions extends RequestInit {
  accessToken?: string;
}

/**
 * Custom fetch wrapper that includes credentials and authentication headers
 * This ensures all API requests include proper authentication
 */
export async function authenticatedFetch(
  url: string,
  options: FetchOptions = {}
): Promise<Response> {
  const { accessToken, headers = {}, ...restOptions } = options;

  // Build full URL if relative path provided
  const fullUrl = url.startsWith('http') ? url : `${API_CONFIG.baseUrl}${url}`;

  // Prepare headers with authentication
  const fetchHeaders: HeadersInit = {
    ...headers,
  };

  // Only set Content-Type if not already set and if we have a body
  // This allows requests like file uploads to set their own Content-Type
  if (!fetchHeaders['Content-Type'] && restOptions.body) {
    // Only set to JSON if body is a string (assumed to be JSON.stringify'd)
    if (typeof restOptions.body === 'string') {
      fetchHeaders['Content-Type'] = 'application/json';
    }
  }

  // Add Authorization header if access token is provided
  if (accessToken) {
    fetchHeaders['Authorization'] = `Bearer ${accessToken}`;
  }

  // Make request with credentials included
  // credentials: 'include' ensures cookies are sent with cross-origin requests
  const response = await fetch(fullUrl, {
    ...restOptions,
    headers: fetchHeaders,
    credentials: 'include', // Important: includes credentials in CORS requests
  });

  return response;
}

/**
 * Helper function to make GET requests
 */
export async function apiGet(
  url: string,
  accessToken?: string
): Promise<Response> {
  return authenticatedFetch(url, {
    method: 'GET',
    accessToken,
  });
}

/**
 * Helper function to make POST requests
 */
export async function apiPost(
  url: string,
  data: unknown,
  accessToken?: string
): Promise<Response> {
  return authenticatedFetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    accessToken,
  });
}

/**
 * Helper function to make PUT requests
 */
export async function apiPut(
  url: string,
  data: unknown,
  accessToken?: string
): Promise<Response> {
  return authenticatedFetch(url, {
    method: 'PUT',
    body: JSON.stringify(data),
    accessToken,
  });
}

/**
 * Helper function to make PATCH requests
 */
export async function apiPatch(
  url: string,
  data: unknown,
  accessToken?: string
): Promise<Response> {
  return authenticatedFetch(url, {
    method: 'PATCH',
    body: JSON.stringify(data),
    accessToken,
  });
}

/**
 * Helper function to make DELETE requests
 */
export async function apiDelete(
  url: string,
  accessToken?: string
): Promise<Response> {
  return authenticatedFetch(url, {
    method: 'DELETE',
    accessToken,
  });
}
